[core] Bump monorepo

Janpot opened the pull request:
Closes #14084.

Just realized I made a few breaking changes in the eslint setup.

@LukasTy Just have two errors on nested imports from the pickers package. How would you prefer to deal with those? just Ignore? Disable in config? Or move the export up one level?

LukasTy commented on docs/src/modules/components/overview/mainDemo/DigitalClock.tsx:
```diff
@@ -14,6 +14,7 @@ import {
 } from '@mui/x-date-pickers/PickersLayout';
 import { renderMultiSectionDigitalClockTimeView } from '@mui/x-date-pickers/timeViewRenderers';
 import { TimeView } from '@mui/x-date-pickers/models';
+// eslint-disable-next-line no-restricted-imports
 import { TimePickerViewRenderers } from '@mui/x-date-pickers/TimePicker/shared';
```
This was clearly left by mistake (ESLint not working correctly 🙈 )
Could we re-export the type from the index and update the import to `@mui/x-date-pickers/TimePicker`?

LukasTy commented on docs/src/modules/components/PickersPlayground.tsx:
```diff
@@ -16,6 +16,7 @@ import MenuItem from '@mui/material/MenuItem';
 import Divider from '@mui/material/Divider';
 import Tooltip from '@mui/material/Tooltip';
 import Typography from '@mui/material/Typography';
+// eslint-disable-next-line no-restricted-imports
```
We don't intend this to be used in production and have it in most of the Pickers demos (docs/data/date-pickers/**)
If it's possible, I would vote to add an exception for this import. 🤔

Janpot replied:
👀 For some reason eslint only reports this instance

alexfauquette commented:
I updated the PR to

- include the fix to have working codesandboxe demonstrations with multiple files https://github.com/mui/material-ui/pull/42939
- I fixed the link check that is only on mui-x. Others should be done in https://github.com/mui/material-ui/pull/43144

LukasTy commented on docs/src/modules/components/overview/mainDemo/DigitalClock.tsx:
```diff
@@ -58,7 +57,7 @@ export default function DigitalClock() {
               minutes: renderMultiSectionDigitalClockTimeView,
               meridiem: renderMultiSectionDigitalClockTimeView,
-            } as TimePickerViewRenderers<Dayjs, 'meridiem' | TimeView, any>
+            } as any
```
We are only using such a hacky approach on the docs demo, where code is not shown.
I would prefer avoiding public exporting of the used type until there is a clear need for it.
WDYT @noraleonte, does that make sense to you? 🤔

Janpot replied:
Couldn't you do the following?
 
```suggestion
            } as StaticTimePickerProps<Dayjs>['viewRenderers']
```

LukasTy replied:
You are right, we could, but do you think that it matters what we cast to in an example, which is not visible to the end user? 🤔

Janpot replied:
we ourselves could benefit from the better type safety.

LukasTy replied:
Sure, it's clearly better than `as any` and will at least have some safeguards. :)

LukasTy replied:
Fixed. 👌

LukasTy reviewed (approved):
I've removed the introduction of type exporting by removing the internal type usage in documentation page component.

alexfauquette commented on docs/data/date-pickers/lifecycle/lifecycle.md:
```diff
@@ -141,7 +141,7 @@ You can find more information [in the dedicated doc section](/x/react-date-picke
 The `onChange` callback is called whenever the current value changes.
 
-If you don't want to listen to the intermediary steps, consider using the [`onAccept` prop](/x/react-date-pickers/lifecycle/#lifecycle-on-pickers-onaccept) instead.
+If you don't want to listen to the intermediary steps, consider using the [`onAccept` prop](/x/react-date-pickers/lifecycle/#lifecycle-on-pickers-quot-onaccept-quot) instead.
```
Not exactly. The `link-check` run on the entire docs each time.

The reason it appears here is that the parser was broken. Updating the monorepo brings the fix for the link check.

LukasTy replied:
Thanks for clarifying. 👍

LukasTy commented:
@Janpot Is there anything else we are waiting for this PR to be ready?

Janpot commented:
> Is there anything else we are waiting for this PR to be ready?

I haven't addressed [this](https://github.com/mui/mui-x/pull/14141#discussion_r1709667796.) yet. But I can also look into it separately, doesn't necessarily have to be ablocker.

LukasTy commented:
> I haven't addressed [this](https://github.com/mui/mui-x/pull/14141#discussion_r1709667796) yet. But I can also look into it separately, doesn't necessarily have to be a blocker.

It could be due to a misconfigured `eslint` config on the `mui-x` side. 🙈 
I would vote for ignoring this question, at least for the time being. 👍

Janpot commented:
> It could be due to a misconfigured eslint config on the mui-x side.

It's likely that the X version of this rule is overwriting a default somewhere. It's quite complicated to extend this rule. Core wasn't even doing it correctly within its repo. It's not a huge problem, I saw that X has made a very detailed setup, I'd like to bring some of this infra to core.

Janpot commented:
Ok, I touched up the `no-retsricted-import` rule and this brought up a few issues. Fixed most of them. I added exceptions for:
* `@mui/x-date-pickers/internals/demo`
* `@mui/x-tree-view/hooks/useTreeViewApiRef`
* `@mui/material/ButtonBase/TouchRipple`: This will definitely need to be fixed, it'll break once we move to package exports

Found an issue where `@mui/monorepo/docs/...` was imported instead of the alias `docs/...` and found that the alias was incorrectly defined. that's why `// @ts-ignore` was used everywhere this was imported. This is fixed now as well

LukasTy commented:
Nice, thanks @Janpot, looks great! 👏 💙

LukasTy commented:
> But I'm not sure if it should be fixed on the core repository, or if tsconfig should point to the monorepo on X side

Without a too-deep investigation, I think we should add this path to the `types`  in the `tsconfig` on the X side.

alexfauquette commented:
I tried ` "@types/react-docgen": ["./node_modules/@mui/monorepo/packages/react-docgen-types"],` in multiple places without success

Janpot commented:
I don't think tsconfig paths apply to anything else than your code, so it's in code that we will need to reference this `@types/react-docgen` that we supplied in the paths. Here I'm doing it in a `docs-env.d.ts` that I introduce to the compilation in the tsconfig `include`. There are probably other ways.
Also pulled in some fixes from PR https://github.com/mui/material-ui/pull/43285. To be merged first.

Janpot commented on docs/data/charts/pie-demo/OnSeriesItemClick.js:
```diff
@@ -1,7 +1,8 @@
 import * as React from 'react';
 import { PieChart } from '@mui/x-charts/PieChart';
 
```
Can't remove this newline for some reason. But didn't take the time to look into it.

LukasTy replied:
This behavior is known and we tend to ignore it. 🙈 
P.S. I take it that adding `type` after `import` doesn't change anything? 🤔

Janpot replied:
It signals that this import is not usable as a javascript value. Can be important depending on the context this is compiled in.

Janpot commented:
Should be ready to review. Have a small [comment](https://github.com/mui/mui-x/pull/14141#discussion_r1716430681) that we could still resolve. I'm a bit allergic to `as any` 😄

LukasTy reviewed (approved):
LGTM!
Thank you for the effort once again. 🙏
